jetty-core/jetty-server/src/main/java/org/eclipse/jetty/server/ResourceService.java

```diff
@@ -365,8 +365,10 @@ protected boolean passConditionalHeaders(Request request, Response response, Htt
                         String matched = matchesEtag(etag, ifnm);
                         if (matched != null)
                         {
+                            // If we do sendError it clears the ETag header.
+                            response.setStatus(HttpStatus.NOT_MODIFIED_304);
                             response.getHeaders().put(HttpHeader.ETAG, matched);
```

Comment:
The header should not be removed, we have tests checking that the ETag header is received in the response. The sendError eventually does a resetContent on the response which clears out specific headers including ETag. 

In Jetty 11 we were setting the status, setting the header and flushing the buffer, so this never went through sendError at all. But I can revert my change and put specific logic in `ServletContextResponse.resetContent()` so this can still go through sendError and keep the ETag header.